beacon_node/lighthouse_network/src/rpc/handler.rs

```diff
@@ -837,6 +837,20 @@ where
             ConnectionEvent::DialUpgradeError(DialUpgradeError { info, error }) => {
                 self.on_dial_upgrade_error(info, error)
             }
+            ConnectionEvent::ListenUpgradeError(e) => {
+                if matches!(e.error, RPCError::InvalidData(_)) {
+                    // Peer is not complying with the protocol. Notify the application and disconnect.
+                    let inbound_substream_id = self.current_inbound_substream_id;
+                    self.current_inbound_substream_id.0 += 1;
+
+                    self.events_out.push(HandlerEvent::Err(HandlerErr::Inbound {
+                        id: inbound_substream_id,
+                        proto: Protocol::DataColumnsByRange, // FIXME: replace this hardcoded protocol
+                        error: e.error,
+                    }));
+                    self.shutdown(None);
+                }
+            }
```

Comment:
> // FIXME: replace this hardcoded protocol

@jxs 

The `RPCError::InvalidData` error is emitted by the inbound codec [here](https://github.com/sigp/lighthouse/blob/70194dfc6a3f4d10c9059610f889ff5a4e863a6a/beacon_node/lighthouse_network/src/rpc/codec.rs#L166-L171). The problem I'm facing is that the handler cannot determine which  protocol the the invalid data belongs to. I think changing the type of [`InboundUpgrade::Error`](https://github.com/sigp/lighthouse/blob/70194dfc6a3f4d10c9059610f889ff5a4e863a6a/beacon_node/lighthouse_network/src/rpc/protocol.rs#L664) from `RPCError` to `(Protocolid, RPCError)` might solves this. What do you think?

Reply:
Updated in https://github.com/sigp/lighthouse/pull/6986/commits/2ff8cf2c123488fd1b3b300df410d25529350264